concent_api/core/payments/bankster.py

```diff
@@ -238,46 +201,47 @@ def finalize_payment(deposit_claim: DepositClaim) -> Optional[str]:
-                        value=deposit_claim.amount,
-                        subtask_id=deposit_claim.subtask_id,
-                    )
-                elif task_to_compute.provider_ethereum_address == deposit_claim.payer_deposit_account.ethereum_address:
-                    ethereum_transaction_hash = service.cover_additional_verification_cost(  # pylint: disable=no-value-for-parameter
-                        provider_eth_address=deposit_claim.payer_deposit_account.ethereum_address,
-                        value=deposit_claim.amount,
-                        subtask_id=deposit_claim.subtask_id,
-                    )
-                else:
-                    assert False
-        else:
-            assert False
+    # If the DepositClaim still exists at this point, Bankster uses SCI to create an Ethereum transaction.
+    if deposit_claim.concent_use_case == ConcentUseCase.FORCED_ACCEPTANCE:
+        ethereum_transaction_hash = service.force_subtask_payment(  # pylint: disable=no-value-for-parameter
+            requestor_eth_address=deposit_claim.payer_deposit_account.ethereum_address,
+            provider_eth_address=deposit_claim.payee_ethereum_address,
+            value=deposit_claim.amount,
+            subtask_id=deposit_claim.subtask_id,
+        )
+    elif deposit_claim.concent_use_case == ConcentUseCase.ADDITIONAL_VERIFICATION:
+        subtask = Subtask.objects.filter(subtask_id=deposit_claim.subtask_id).first()  # pylint: disable=no-member
+        if subtask is not None:
+            task_to_compute = deserialize_message(subtask.task_to_compute.data.tobytes())
+            if task_to_compute.requestor_ethereum_address == deposit_claim.payer_deposit_account.ethereum_address:
+                ethereum_transaction_hash = service.force_subtask_payment(  # pylint: disable=no-value-for-parameter
+                    requestor_eth_address=deposit_claim.payer_deposit_account.ethereum_address,
+                    provider_eth_address=deposit_claim.payee_ethereum_address,
+                    value=deposit_claim.amount,
+                    subtask_id=deposit_claim.subtask_id,
+                )
+            elif task_to_compute.provider_ethereum_address == deposit_claim.payer_deposit_account.ethereum_address:
+                ethereum_transaction_hash = service.cover_additional_verification_cost(  # pylint: disable=no-value-for-parameter
+                    provider_eth_address=deposit_claim.payer_deposit_account.ethereum_address,
+                    value=deposit_claim.amount,
+                    subtask_id=deposit_claim.subtask_id,
+                )
+            else:
+                assert False
+    else:
+        assert False
 
+    with transaction.atomic(using='control'):
         # Bankster puts transaction ID in DepositClaim.tx_hash.
         ethereum_transaction_hash = adjust_transaction_hash(ethereum_transaction_hash)
         deposit_claim.tx_hash = ethereum_transaction_hash
```

Comment:
Since `deposit_claim` was fetched in another transaction, it's no longer safe to use it. We've already released our lock and some other process could have modified it and we risk overwriting his changes.

That's the theory. In practice we won't ever run this function on the same claim in parallel (see https://github.com/golemfactory/concent/pull/1041/files#r242920641) but that's not because the function is correct but because of how Concent uses it. So we can keep using the object - but again, please add a comment explaining that this is correct only with that assumption. If I saw it without knowing about this assumption I would naturally think it was a bug.